rust/src/pgsql/pgsql.rs

```diff
@@ -0,0 +1,853 @@
+                Some(PgsqlStateProgress::SimpleAuthenticationReceived)
+            }
+            PgsqlBEMessage::RowDescription(_) => Some(PgsqlStateProgress::DataRowReceived),
+            PgsqlBEMessage::CommandComplete(_) => {
+                // TODO Do we want to compare the command that was stored when
+                // query was sent with what we received here?
+                Some(PgsqlStateProgress::CommandCompletedReceived)
+            }
+            PgsqlBEMessage::ErrorResponse(_) => Some(PgsqlStateProgress::ErrorMessageReceived),
+            _ => {
+                // TODO handle unexpected situations here?
+                // We don't always have to change current state when we see a response...
+                None
+            }
+        }
+    }
+
+    fn state_based_resp_parsing(
+        state: PgsqlStateProgress, input: &[u8],
+    ) -> Result<(&[u8], parser::PgsqlBEMessage), nom::Err<(&[u8], nom::error::ErrorKind)>> {
+        if state == PgsqlStateProgress::SSLRequestReceived {
+            parser::parse_ssl_response(input)
+        } else {
+            parser::pgsql_parse_response(input)
+        }
+    }
+
+    fn parse_response(&mut self, input: &[u8], flow: *const Flow) -> AppLayerResult {
+        // We're not interested in empty responses.
+        if input.len() == 0 {
+            return AppLayerResult::ok();
+        }
+
+        if self.response_gap {
+            if !probe_tc(input) {
+                // Out of sync, we'll just try again next time.
+                SCLogDebug!("Suricata interprets there's a gap in the response");
+                return AppLayerResult::ok();
+            }
+
+            // It seems we're in sync with a message header, clear gap state and keep parsing.
+            self.response_gap = false;
+        }
+
+        let mut start = input;
+        while start.len() > 0 {
+            match PgsqlState::state_based_resp_parsing(self.state_progress, start) {
+                Ok((rem, response)) => {
+                    start = rem;
+                    SCLogDebug!("Response is {:?}", &response);
+                    if let Some(state) = self.response_get_next_state(&response, flow) {
+                        self.state_progress = state;
+                    };
+                    if let Some(tx) = self.find_or_create_tx() {
+                        tx.responses.push(response);
+                    } else {
+                        // Not sure if this is the best solution here, almost sure it isn't...
+                        return AppLayerResult::err();
+                    };
+                }
+                Err(nom::Err::Incomplete(_needed)) => {
+                    let consumed = input.len() - start.len();
+                    let needed_estimation = start.len() + 1;
+                    SCLogDebug!(
+                        "Needed: {:?}, estimated needed: {:?}, start is {:?}",
+                        _needed,
```

Comment:
If you are using `_needed`, it probably does not need the `_` in the beginning.

Reply:
I think Rust complained because it was only used in a Debug statement... Or maybe because I wasn't using it and the debug statement came later on. I'll check if it complains in the next version :P

Reply:
Ah right! Debug statements.